### ISSUE TITLE
Redundant `ln_pre`

### ISSUE THREAD

**Issue by carefree0910** [openai/CLIP#152]
First of all, thanks for your great work!

And I found out that the `ln_pre` seems to be a redundant component:

https://github.com/openai/CLIP/blob/3b473b0e682c091a9e53623eebc1ca1657385717/clip/model.py#L212

Because you've already applied `pre_norm` mechanism in `ResidualAttentionBlock`.

I'm just wondering whether this is intended or some kinds of 'typo', thanks in advance!

**Comment by jongwook** [comment]
I think they're not completely redundant because the 'identity' connection in `ResidualAttentionBlock` does not go through layernorms.

**Comment by russelldc** [comment]
I'm not sure this is relevant to this specific issue, but I found a significant difference while using CLIP to guide image generation.
I tried removing the use of that ln_pre in the VisionTransformer's forward(). It was no longer able to guide the generation to anything other than red/green stripes.

**Comment by carefree0910** [comment]
> I think they're not completely redundant because the 'identity' connection in `ResidualAttentionBlock` does not go through layernorms.

I agree with you that it has impact on the model, but I'm not sure whether this is 'intended', because other main stream implementations don't have this kind of `ln_pre` before the 'identity' connection 😉

**Comment by carefree0910** [comment]
> I'm not sure this is relevant to this specific issue, but I found a significant difference while using CLIP to guide image generation.
> I tried removing the use of that ln_pre in the VisionTransformer's forward(). It was no longer able to guide the generation to anything other than red/green stripes.

Not sure if I get you right, but when I say 'redundant' I mean maybe we can re-train the model without it.

If you are using a pre-trained model, the `ln_pre` is definitely useful because the pre-trained weights are trained based on it 😆

**Comment by jongwook** [comment]
We had it since when we initially iterated on transformer language model where it had ResNet features as prefix (LM RN50 in paper), and it got kinda stuck in the code for the contrastive version. I agree with you that it's not particularly necessary in the text encoder when training from scratch.
